rust/codec-jats/src/decode/front.rs

```diff
@@ -1,16 +1,331 @@
+        };
+    }
+}
+
+/// Decode a `<subj-group>` element by adding the text content of the inner `<subject>`
+/// to the article's `about` property
+fn decode_subj_group(path: &str, node: &Node, article: &mut Article, losses: &mut Losses) {
+    let subject_type = node.attribute("subj-group-type").map(String::from);
+
+    record_attrs_lost(path, node, ["subj-group-type"], losses);
+
+    let mut subject = None;
+    for child in node.children() {
+        if child.tag_name().name() == "subject" {
+            if let Some(text) = child.text() {
+                subject = Some(text.to_string());
+            }
+        } else {
+            record_node_lost(path, &child, losses);
+        }
+    }
+    let Some(subject) = subject else { return };
+
+    let item = ThingType::PropertyValue(PropertyValue {
+        property_id: subject_type,
+        value: Primitive::String(subject),
+        ..Default::default()
+    });
+
+    if let Some(about) = &mut article.options.about {
+        about.push(item);
+    } else {
+        article.options.about = Some(vec![item])
+    }
+}
+
+/// Decode an `<article-id>` element
+fn decode_article_id(path: &str, node: &Node, article: &mut Article, losses: &mut Losses) {
+    let property_id = node.attribute("pub-id-type").map(String::from);
+
+    record_attrs_lost(path, node, ["pub-id-type"], losses);
+
+    let Some(id) = node.text() else { return };
+
+    let item = if property_id.is_some() {
+        PropertyValueOrString::PropertyValue(PropertyValue {
+            property_id,
+            value: Primitive::String(id.into()),
+            ..Default::default()
+        })
+    } else {
+        PropertyValueOrString::String(id.into())
+    };
+
+    if let Some(about) = &mut article.options.identifiers {
+        about.push(item);
+    } else {
+        article.options.identifiers = Some(vec![item])
+    }
+}
+
+/// Decode an `<article-version>` element
+fn decode_article_version(path: &str, node: &Node, article: &mut Article, losses: &mut Losses) {
+    record_attrs_lost(path, node, [], losses);
+
+    if let Some(version) = node.text() {
+        article.options.version = Some(StringOrNumber::String(version.into()))
+    };
+}
+
+///Decode an `<pub-date>` element
+fn decode_pub_date(path: &str, node: &Node, article: &mut Article, losses: &mut Losses) {
+    record_attrs_lost(path, node, [], losses);
+    let mut day = "";
+    let mut month = "";
+    let mut year = "";
+
+    for child in node.children() {
+        let tag = child.tag_name().name();
+        if tag == "day" {
+            day = child.text().unwrap_or_default()
+        } else if tag == "month" {
+            month = child.text().unwrap_or_default()
+        } else if tag == "year" {
+            year = child.text().unwrap_or_default()
+        }
+    }
+    article.date_published = Some(Date {
+        value: [year, "-", month, "-", day].concat(),
+        ..Default::default()
+    });
```

Comment:
This would have given invalid date strings if elements were missing or empty so I fixed it in 1250e3616239bfe6360eec012c3395486e49f6eb